Update from Upstream

Alchez opened the pull request.

Commit:
fix: changing percent field precision to global default

(cherry picked from commit 729ee6ea9ea1138043a501632891cfd03fc68396)

Commit:
fix: sider issue fixed

(cherry picked from commit 8aeb1716642a1e275ddbc83c77501a24529dd1c5)

Commit:
fix: fallback for global default precision

(cherry picked from commit f0ae807df83009fc2fe9f7e82111b3404a395f60)

Commit:
fix: fallback for percent field precision to 2

(cherry picked from commit e7417ebe486bd5d60a73056735743c079fc44597)

gavindsouza committed:
style: Re-format and use const instead of var

(cherry picked from commit 8ef2f7ed1896f7a055e42b456bba8152a69ac1a7)

surajshetty3416 committed:
fix: Raise PageDoesNotExistError instead of DoesNotExistError (#12485)

- to avoid it getting mixed up with Document Not Found error

(cherry picked from commit 71dd300ae300aac8e00b2c1bc79591bfe36c8f41)

gavindsouza committed:
fix: Remove reloading statements for non existing dts

fahimalizain committed:
fix: frappe.local.lang resolution

gavindsouza committed:
perf: Skip guess_language if _lang is provided

barredterra committed:
fix: translation of email salutation

(cherry picked from commit 23231fcbba400a0ad0af7b2f1cf9a5e10c49f83e)

pateljannat committed:
Merge pull request #13517 from frappe/mergify/bp/version-12-hotfix/pr-12010

fix: changing percent field precision to global default (backport #12010)

dependabot-preview[bot] committed:
chore(deps): [security] bump urllib3 from 1.25.8 to 1.25.9

Bumps [urllib3](https://github.com/urllib3/urllib3) from 1.25.8 to 1.25.9. **This update includes a security fix.**
- [Release notes](https://github.com/urllib3/urllib3/releases)
- [Changelog](https://github.com/urllib3/urllib3/blob/main/CHANGES.rst)
- [Commits](https://github.com/urllib3/urllib3/compare/1.25.8...1.25.9)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

surajshetty3416 committed:
Merge pull request #13525 from frappe/dependabot/pip/version-12-hotfix/urllib3-1.25.9

dependabot-preview[bot] committed:
chore(deps): [security] bump color-string from 1.5.3 to 1.5.5

Bumps [color-string](https://github.com/Qix-/color-string) from 1.5.3 to 1.5.5. **This update includes a security fix.**
- [Release notes](https://github.com/Qix-/color-string/releases)
- [Changelog](https://github.com/Qix-/color-string/blob/master/CHANGELOG.md)
- [Commits](https://github.com/Qix-/color-string/commits/1.5.5)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

gavindsouza committed:
Merge pull request #13552 from frappe/dependabot/npm_and_yarn/version-12-hotfix/color-string-1.5.5

chore(deps): [security] bump color-string from 1.5.3 to 1.5.5

leela committed:
fix: Email signature not showing up

leela committed:
Merge pull request #13562 from leela/fix-email-signature

fix: Email signature not showing up

nextchamp-saqib committed:
feat: return promise while reloading doc (#13572)

nabinhait committed:
Merge branch 'version-12' into version-12-hotfix

gavindsouza committed:
fix: Set currency only if df exists

Co-authored-by: marination <maricadsouza221197@gmail.com>

gavindsouza committed:
Merge pull request #13595 from gavindsouza/print-format-df

fix: Set currency only if df exists

surajshetty3416 committed:
Merge pull request #13604 from frappe/mergify/bp/version-12-hotfix/pr-13503

fix: translation of email salutation (backport #13503)

gavindsouza committed:
fix: Hide "Attach File" in form if max limit reached

* Correct element identifier
* Fix visibility toggler

gavindsouza committed:
Merge pull request #13618 from frappe/hide-attachment-cond

fix: Hide "Attach File" in form if max limit reached

mituldavid committed:
fix: Accurately cast fieldtype in frappe.db.get_single_value()

(cherry picked from commit a1e84d234108c950d2d00a3df0eb70247c154966)

mituldavid committed:
refactor(minor): Delete redundant imports

(cherry picked from commit 64840aac6bcca0e1af5203a0d3e005359ade0f1b)

mituldavid committed:
test: Add test for frappe.db.get_single_value

(cherry picked from commit 1e839f0ab1ca0e00a0492a2f67fcb7eab13a6720)

# Conflicts:
#	frappe/tests/test_db.py

mituldavid committed:
refactor(minor): Fix indentation

(cherry picked from commit 95550861a74eb1d56b72ef7536e111b703b2fa86)

fahimalizain committed:
fix: preferred_language cookie support for all users

nabinhait committed:
Merge branch 'version-12-hotfix' into v12-pre-release

mituldavid committed:
refactor: Improve readability, consistency

(cherry picked from commit 934fc17a3b0aa10c65cd24e87606c674070c61f7)

gavindsouza committed:
fix: Resolve conflicts & imports

gavindsouza committed:
Merge pull request #13683 from frappe/mergify/bp/version-12-hotfix/pr-13655

fix: Accurately cast fieldtype in frappe.db.get_single_value() (backport #13655)

gavindsouza committed:
refactor: Simplify HTTPRequest class

* For the sake of Readability and ease of understanding
* Style updates

dependabot-preview[bot] committed:
chore(deps): [security] bump urllib3 from 1.25.9 to 1.26.5

Bumps [urllib3](https://github.com/urllib3/urllib3) from 1.25.9 to 1.26.5. **This update includes a security fix.**
- [Release notes](https://github.com/urllib3/urllib3/releases)
- [Changelog](https://github.com/urllib3/urllib3/blob/main/CHANGES.rst)
- [Commits](https://github.com/urllib3/urllib3/compare/1.25.9...1.26.5)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

surajshetty3416 committed:
fix: Add missing fieldtypes in for custom fields

surajshetty3416 committed:
fix: Add missing fieldtypes in for custom fields

gavindsouza committed:
refactor: Set Language in HTTPHeader

Order of priority for setting language:
1. Form Dict => _lang
2. Cookie => preferred_language
3. Request Header => Accept-Language
4. User document => language
5. System Settings => language

Cookie is placed at #2 since the language picker in the navbar depends
on it. And the Accept-Language header sends values based on the client's locales.

---

Form Dict _lang now accepts language codes too. Previously, language
names were used...for whatever reason.

gavindsouza committed:
refactor: Rename guess_language as get_language

Guess suggests there's some AI involvement. The get_language function
has a defined priority. It is deterministic, hence teh name change.

sagarvora committed:
fix: escape quotes before declaring variables when making a new app

(cherry picked from commit b23dd711162ff03922c7469fd478623bad45dff1)

gavindsouza committed:
fix: Don't redefine datetime

* Sort imports
* Update file header

gavindsouza committed:
fix: Clear preferred_language cookie post login

If preferred_language was set in cookie pre login, clear it after a
successful login so that User or Site specific settings can be applied

gavindsouza committed:
ci: Which urllib version is installed anyway?

Had to test it this way since couldn't install py3.6 on any new
servers/my machines. This is easier rather than setting up a new server
;)

gavindsouza committed:
Revert "ci: Which urllib version is installed anyway?"

This reverts commit 4529f083356a31cebec7881f76fb31f7745edbfe.

gavindsouza committed:
build: Pin urllib to 1.25.11 instead

Seems like the latest compatible version for 3.6. PY2 works fine with
1.26.5 though

gavindsouza committed:
Merge pull request #13551 from frappe/dependabot/pip/version-12-hotfix/urllib3-1.26.5

chore(deps): [security] [v12] bump urllib3 from 1.25.9 to 1.25.11

surajshetty3416 committed:
fix: Do not create energy points for disabled users

(cherry picked from commit bd854bb368023c43a90d1cb044a5282a082e9371)

# Conflicts:
#	frappe/core/doctype/user/user.py
#	frappe/social/doctype/energy_point_rule/energy_point_rule.py

gavindsouza committed:
refactor(minor): frappe.translate module

* Remove unset limit for lru cache in get_parent_language
* Simplify get_user_lang and add relevant comment

surajshetty3416 committed:
test: Add test case to validate EP for disabled user

(cherry picked from commit bd91a33bb91ee57a3b6b73ea7cda1c285e81a0e4)

surajshetty3416 committed:
chore: Add change log for v12.20.0

surajshetty3416 committed:
Merge pull request #13714 from surajshetty3416/change-log-v12-20-0

surajshetty3416 committed:
fix: Resolve conflicts

surajshetty3416 committed:
feat: Add has_value_changed method

backported from https://github.com/frappe/frappe/pull/10838

surajshetty3416 committed:
fix: Replace unsupported code

surajshetty3416 committed:
Merge pull request #13709 from frappe/mergify/bp/version-12-hotfix/pr-13708

mergify[bot] committed:
fix: Do not create energy points for disabled users (backport #13708) (backport #13709) (#13717)

gavindsouza committed:
Merge pull request #13700 from frappe/mergify/bp/version-12-hotfix/pr-13692

fix: escape quotes before declaring variables when making a new app (backport #13692)

surajshetty3416 committed:
chore: Update change log

surajshetty3416 committed:
Merge pull request #13727 from surajshetty3416/fix-customize-form-version-pre-release

nabinhait committed:
bumped to version 12.20.0

nabinhait committed:
Merge branch 'v12-pre-release' into version-12

nabinhait committed:
Merge branch 'version-12' into version-12-hotfix

surajshetty3416 committed:
fix: Trigger worklow action on update_after_submit event

(cherry picked from commit b690374afeb17fda8230151370766b5488de054a)

surajshetty3416 committed:
Merge pull request #13725 from surajshetty3416/fix-customize-form-version-12-hotfix

surajshetty3416 committed:
Merge pull request #13737 from frappe/mergify/bp/version-12-hotfix/pr-13734

fix: Trigger workflow action on update_after_submit event (backport #13734)

gavindsouza committed:
fix(guess_language): Make it PY2 compatible

gavindsouza committed:
fix(translate): Add broken import 'codecs'

surajshetty3416 committed:
Merge pull request #13795 from frappe/mergify/bp/version-12-hotfix/pr-12485

gavindsouza committed:
chore: Re-add unicode literals import for PY2

Since v12 is PY2 only ;)

gavindsouza committed:
Merge pull request #13767 from gavindsouza/lang-resolution-v12

refactor: Request Language Resolution

gavindsouza committed:
refactor(lang): frappe.translate.get_language

User.language should be given higher priority in terms of
authenticated user since they chose it. Even higher than the
browser they're using...even if the system locales aren't set
properly and browser isn't configured properly

ankush committed:
perf: fix unneneccsary clearing of db.value_cache

(cherry picked from commit 8ed2c87e00ca26d9439c85c66b840557a1799bd3)

SaiFi0102 committed:
fix: do not reset doc_before_save on db_set

(cherry picked from commit 7400d7ca37a6a57ecf9aa6ad725c1c33c04dc8a1)

surajshetty3416 committed:
Merge pull request #13882 from frappe/mergify/bp/version-12-hotfix/pr-13880

gavindsouza committed:
Merge pull request #13903 from gavindsouza/lang-resolution-loggedin-v12

refactor(frappe.lang): frappe.translate.get_language [v12 port]

shariquerik committed:
fix: Return async function for sequential execution

Co-authored-by: Suraj Shetty <surajshetty3416@gmail.com>

surajshetty3416 committed:
Merge pull request #13913 from shariquerik/bulk-delete-grid-item

surajshetty3416 committed:
Merge pull request #13888 from frappe/mergify/bp/version-12-hotfix/pr-13883

fix: do not reset doc_before_save on db_set (backport #13883)

mergify[bot] committed:
fix: format string for formating date (backport #13878) (#13894)

Co-authored-by: Afshan <afshan13k@gmail.com>
Co-authored-by: Afshan <33727827+AfshanKhan@users.noreply.github.com>

rohitwaghchaure committed:
Merge pull request #13979 from rohitwaghchaure/merge-v12-hotfix-to-pre-release

chore: merge v12 hotfix to v12 pre-release

rohitwaghchaure committed:
chore: change log for v12.21.0

rohitwaghchaure committed:
Merge pull request #13983 from rohitwaghchaure/change-log-v12-21

chore: change log for v12.21.0

rohitwaghchaure committed:
Merge branch 'v12-pre-release' into version-12

rohitwaghchaure committed:
bumped to version 12.21.0

Commit:
Merge remote-tracking branch 'upstream/parsimony-production' into version-12

ghost commented:
### Bug Fixes

* changing percent field precision to global default ([2906e37](https://github.com/ParsimonyGit/frappe/commit/2906e37a96e617685a351078235d5f9734f7f8b6))
* sider issue fixed ([0364b40](https://github.com/ParsimonyGit/frappe/commit/0364b40fd98ccfd9711aa2b4689c4420bba341dc))
* fallback for global default precision ([5a7bf6f](https://github.com/ParsimonyGit/frappe/commit/5a7bf6f24198dfef88fae9894603ec92ab77b8f4))
* fallback for percent field precision to 2 ([25d9be4](https://github.com/ParsimonyGit/frappe/commit/25d9be424be40d04ab3033b52d67a5b5501d7d1e))
* Email signature not showing up ([4cc54c2](https://github.com/ParsimonyGit/frappe/commit/4cc54c271045c8f1f38e7a68d0c8964f8b8e70ec))
* Set currency only if df exists ([3584503](https://github.com/ParsimonyGit/frappe/commit/3584503574ed2345da9e67d94b81ab89d1f1085e))
* translation of email salutation ([d214fdf](https://github.com/ParsimonyGit/frappe/commit/d214fdf8b6d8c681aa34714bbf54ec8d23c7d5b8))
* Hide "Attach File" in form if max limit reached ([f3b5981](https://github.com/ParsimonyGit/frappe/commit/f3b5981bf808b420a293fb89307d48b9371db34d))
* Accurately cast fieldtype in frappe.db.get_single_value() ([dbc6ff1](https://github.com/ParsimonyGit/frappe/commit/dbc6ff17ccfdfcfb37943baff161d0ae488210c4))
* Resolve conflicts & imports ([d36ae38](https://github.com/ParsimonyGit/frappe/commit/d36ae3840d750d6bce70d9efdf1895e63bd43413))
* escape quotes before declaring variables when making a new app ([59d53a6](https://github.com/ParsimonyGit/frappe/commit/59d53a6de6f1f33a4b354f84cb5b501cd9f490bf))


### Styles

* Re-format and use const instead of var ([babd9d1](https://github.com/ParsimonyGit/frappe/commit/babd9d11c613b427d4a727bec736d3cebc4e9f86))


### Chore

*  **deps:** [security] bump urllib3 from 1.25.8 to 1.25.9 ([d8c1776](https://github.com/ParsimonyGit/frappe/commit/d8c1776195569a6d581d8b1514b41586039477e2))
*  **deps:** [security] bump color-string from 1.5.3 to 1.5.5 ([8d80107](https://github.com/ParsimonyGit/frappe/commit/8d80107214dd373edb6f8f67f15b97ffa803c2c9))
*  **deps:** [security] bump urllib3 from 1.25.9 to 1.26.5 ([6cffd46](https://github.com/ParsimonyGit/frappe/commit/6cffd46315930e54e73b57f125d4e41db8133eea))


### Features

* return promise while reloading doc (#13572) ([79bfafe](https://github.com/ParsimonyGit/frappe/commit/79bfafed7da1b73d172bb6f7008885270be2c194))


### Code Refactoring

*  **minor:** Delete redundant imports ([18228b3](https://github.com/ParsimonyGit/frappe/commit/18228b3c238721302ee44c24e867982d77becc2a))
*  **minor:** Fix indentation ([51feb45](https://github.com/ParsimonyGit/frappe/commit/51feb4530ae9399f87f24770e15814295cc367a7))
* Improve readability, consistency ([74041d2](https://github.com/ParsimonyGit/frappe/commit/74041d222386d106c34676e5f5ca72b4d1342c8b))


### Tests

* Add test for frappe.db.get_single_value ([7d3a542](https://github.com/ParsimonyGit/frappe/commit/7d3a542deae217bac5258d4086290bc54eeb56bf))


### Contributors

<a href="https://github.com/gavindsouza">gavindsouza</a>, <a href="https://github.com/pateljannat">pateljannat</a>, <a href="https://github.com/dependabot-preview">dependabot-preview[bot]</a>, <a href="https://github.com/surajshetty3416">surajshetty3416</a>, <a href="https://github.com/leela">leela</a>, <a href="https://github.com/nextchamp-saqib">nextchamp-saqib</a>, <a href="https://github.com/nabinhait">nabinhait</a>, <a href="https://github.com/barredterra">barredterra</a>, <a href="https://github.com/mituldavid">mituldavid</a>, <a href="https://github.com/sagarvora">sagarvora</a>

<details>
<summary>Commit-Lint commands</summary>
<br />

You can trigger Commit-Lint actions by commenting on this PR:
- `@Commit-Lint merge patch` will merge dependabot PR on "patch" versions (X.X.Y - Y change)
- `@Commit-Lint merge minor` will merge dependabot PR on "minor" versions (X.Y.Y - Y change)
- `@Commit-Lint merge major` will merge dependabot PR on "major" versions (Y.Y.Y - Y change)
- `@Commit-Lint merge disable` will desactivate merge dependabot PR
- `@Commit-Lint review` will approve dependabot PR
- `@Commit-Lint stop review` will stop approve dependabot PR



</details>

gavindsouza committed:
fix: Define jinja2 exception

Got removed by human error via https://github.com/frappe/frappe/pull/13767

gavindsouza committed:
Merge pull request #14082 from gavindsouza/templateerror-v12

fix(translate): Define jinja2 exception

Commit:
chore: Bump to v12.21.1

gavindsouza committed:
Merge branch 'version-12' into version-12-hotfix

gavindsouza committed:
Merge pull request #14089 from frappe/version-12-hotfix

chore: Merge version-12-hotfix into v12-pre-release

gavindsouza committed:
Merge pull request #14091 from frappe/v12-pre-release

chore: Merge v12-pre-release into version-12

gavindsouza committed:
Revert "fix: Accurately cast fieldtype in frappe.db.get_single_value() (backport #13655)"

gavindsouza committed:
Merge pull request #14117 from frappe/revert-13683-mergify/bp/version-12-hotfix/pr-13655

Revert "fix: Accurately cast fieldtype in frappe.db.get_single_value() (backport #13655)"

vamagithub committed:
fix: Problems while using translations via Globe Symbol

Co-authored-by: Vama Mehta <vama.mehta@inqubit.in>

gavindsouza committed:
feat: @frappe.write_only

When would you want to use it? Typically if you want to be sure that
"this" particular thing will be make WRITES to the database. If this
were to be called via a function wrapped with @frappe.read_only. This
will momentarily switch the connection from replica to primary

(cherry picked from commit 61d067710ffd1819f0ed26548ea64d915a8cbc45)

gavindsouza committed:
fix: Make `make_access_log` write only

Write to primary db if connection is currently set to replica

(cherry picked from commit 70ac05025b274d70cd6f63e4ef165ac985c0b234)

gavindsouza committed:
fix: Get rid of pointless except: raise

(cherry picked from commit 309eaabcf96f35906e0a16a4ba35ccb3049e055b)

surajshetty3416 committed:
Merge pull request #14146 from frappe/mergify/bp/version-12-hotfix/pr-14142

lukptr committed:
feat: Able to use custom port for database load balancers

(cherry picked from commit 85834d7a000759af5526c7b3f93d04b44bf6bc88)

lukptr committed:
feat: custom database port for read-only replica configuration

(cherry picked from commit 1ee23859447834f166bc4dbdd05cbd2e7c8e510a)

surajshetty3416 committed:
fix(translation): Set translated text as html to get unescaped output

(cherry picked from commit 268d344f1a5b544ae4daba539147a170065d1299)

# Conflicts:
#	frappe/templates/includes/login/login.js

surajshetty3416 committed:
fix: Resolve conflicts

surajshetty3416 committed:
Merge pull request #14221 from frappe/mergify/bp/version-12-hotfix/pr-14214

fix(translation): Set translated text as html to get unescaped output (backport #14214)

mergify[bot] committed:
fix(grid): date or datetime fields disappearing on save (#14227)

(cherry picked from commit fb508607e0c45dc776a41c2fd661a145a751f9c0)

Co-authored-by: Sagar Vora <sagar@resilient.tech>

surajshetty3416 committed:
Merge pull request #14257 from frappe/mergify/bp/version-12-hotfix/pr-14159

feat: custom database port for read-only replica configuration (backport #14159)

rohitwaghchaure committed:
Merge branch 'version-12-hotfix' into merge-v12-hotfix-pre-release-for-v12-21-2

rohitwaghchaure committed:
Merge pull request #14259 from rohitwaghchaure/merge-v12-hotfix-pre-release-for-v12-21-2

chore: merge version-12-hotfix into version-12-pre-release

rohitwaghchaure committed:
chore: change log for v12.22.0

rohitwaghchaure committed:
Merge pull request #14261 from rohitwaghchaure/added-change-log-v12-22-0

chore: change log for v12.22.0

rohitwaghchaure committed:
bumped to version 12.22.0

rohitwaghchaure committed:
Merge branch 'v12-pre-release' into version-12

mergify[bot] committed:
fix: properly queue tasks in `frappe.model.trigger` (#14264)

(cherry picked from commit 778ab5919c9b1752dccf01ebe513af4a269a69e4)

Co-authored-by: Sagar Vora <sagar@resilient.tech>

Commit:
Merge remote-tracking branch 'frappe/version-12' into version-12

gavindsouza committed:
Merge pull request #14293 from ankush/patch_fixe

fix: broken v10-v12 migration

gavindsouza committed:
Merge pull request #14306 from frappe/version-12-hotfix

chore: merge version-12-hotfix into version-12-pre-release

Commit:
chore: Bump to vRelease 12.22.1

gavindsouza committed:
Merge pull request #14308 from frappe/v12-pre-release

chore: Merge v12-pre-release into version-12

Commit:
Merge remote-tracking branch 'frappe/version-12' into version-12

gavindsouza committed:
chore: Update and fix __version__ tag

Commit:
Merge remote-tracking branch 'frappe/version-12' into version-12
